phpBB/phpbb/textformatter/s9e/parser.php

```diff
@@ -0,0 +1,355 @@
+	* @param  factory $factory
+	* @return null
+	*/
+	public function __construct(\phpbb\cache\driver\driver_interface $cache, $key, \phpbb\user $user, factory $factory)
+	{
+		$this->user = $user;
+
+		$parser = $cache->get($key);
+		if (!$parser)
+		{
+			$objects = $factory->regenerate();
+			$parser  = $objects['parser'];
+		}
+
+		$this->parser = $parser;
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function parse($text)
+	{
+		return $this->parser->parse($text);
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function disable_bbcode($name)
+	{
+		$this->parser->disableTag(strtoupper($name));
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function disable_bbcodes()
+	{
+		$this->parser->disablePlugin('BBCodes');
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function disable_censor()
+	{
+		$this->parser->disablePlugin('Censor');
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function disable_magic_url()
+	{
+		$this->parser->disablePlugin('Autoemail');
+		$this->parser->disablePlugin('Autolink');
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function disable_smilies()
+	{
+		$this->parser->disablePlugin('Emoticons');
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function enable_bbcode($name)
+	{
+		$this->parser->enableTag(strtoupper($name));
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function enable_bbcodes()
+	{
+		$this->parser->enablePlugin('BBCodes');
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function enable_censor()
+	{
+		$this->parser->enablePlugin('Censor');
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function enable_magic_url()
+	{
+		$this->parser->enablePlugin('Autoemail');
+		$this->parser->enablePlugin('Autolink');
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function enable_smilies()
+	{
+		$this->parser->enablePlugin('Emoticons');
+	}
+
+	/**
+	* {@inheritdoc}
+	*
+	* This will translate the log entries found in s9e\TextFormatter's logger into phpBB error
+	* messages
+	*/
+	public function get_errors()
+	{
+		$errors = array();
+
+		foreach ($this->parser->getLogger()->get() as $entry)
+		{
+			list($type, $msg, $context) = $entry;
+
+			if ($msg === 'Tag limit exceeded')
+			{
+				if ($context['tagName'] === 'E')
+				{
+					$errors[] = $this->user->lang('TOO_MANY_SMILIES', $context['tagLimit']);
+				}
+				else if ($context['tagName'] === 'URL')
+				{
+					$errors[] = $this->user->lang('TOO_MANY_URLS', $context['tagLimit']);
+				}
+			}
+			else if ($msg === 'MAX_FONT_SIZE_EXCEEDED')
+			{
+				$errors[] = $this->user->lang($msg, $context['max_size']);
+			}
+			else if (preg_match('/^MAX_(?:FLASH|IMG)_(HEIGHT|WIDTH)_EXCEEDED$/D', $msg, $m))
+			{
+				$errors[] = $this->user->lang($msg, $context['max_' . strtolower($m[1])]);
+			}
+			else if ($msg === 'Tag is disabled')
+			{
+				$name = strtolower($context['tag']->getName());
+				$errors[] = $this->user->lang('UNAUTHORISED_BBCODE', '[' . $name . ']');
+			}
+			else if ($msg === 'UNABLE_GET_IMAGE_SIZE')
+			{
+				$errors[] = $this->user->lang[$msg];
```

Comment:
Array access is kind of legacy I think. Using a method call is definitively better and anyway I don't think the lang array will stay public forever